### PR TITLE
Add support for basic certificate verification policies in C++

### PR DESCRIPTION
**Pull request by vekterli**
@havardpe and @bjorncs please review

Extends TLS config JSON file with an `allowed-peers` object, which
if non-empty specifies a set of policies that a peer may match. If
at least one policy exists a peer must match all requirements in
any single policy to be allowed to connect. I.e. it's sufficient
to match 1 policy out of many.

### CODE REVIEW

File: vespalib/src/tests/net/tls/policy_checking_certificate_verifier/policy_checking_certificate_verifier_test.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    EXPECT_FALSE(glob_matches("f?o", "fooo"));
+    EXPECT_FALSE(glob_matches("f?o", "ffoo"));
+    EXPECT_FALSE(glob_matches("f?o", "f.o"));
+}
+
+TEST("special basic regex characters are escaped") {
+    EXPECT_TRUE(glob_matches("$[.\\^", "$[.\\^"));
+}
+
+TEST("special extended regex characters are ignored") {
```
Comment (havardpe) [inline vespalib/src/tests/net/tls/policy_checking_certificate_verifier/policy_checking_certificate_verifier_test.cpp]: what is the difference between 'escaped' and 'ignored' ?

File: vespalib/src/tests/net/tls/policy_checking_certificate_verifier/policy_checking_certificate_verifier_test.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+TEST("Specially constructed set of policies allows all authenticated peers") {
+    auto allow_all = AllowedPeers::allow_all_authenticated();
+    EXPECT_TRUE(allow_all.allows_all_authenticated());
+    EXPECT_TRUE(verify(allow_all, creds_with_dns_sans({{"anything.goes"}})));
+}
+
+TEST("Non-empty policies do not allow all unauthenticated peers") {
+    auto allow_not_all = allowed_peers({policy_with({required_san_dns("hello.world")})});
+    EXPECT_FALSE(allow_not_all.allows_all_authenticated());
```
Comment (havardpe) [inline vespalib/src/tests/net/tls/policy_checking_certificate_verifier/policy_checking_certificate_verifier_test.cpp]: the test says both authenticated and unauthenticated...

File: vespalib/src/vespa/vespalib/net/tls/policy_checking_certificate_verifier.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "policy_checking_certificate_verifier.h"
+
+namespace vespalib::net::tls {
+
+namespace {
+
+bool matches_single_san_requirement(const PeerCredentials& peer_creds, const RequiredPeerCredential& requirement) {
+    for (auto& provided_cred : peer_creds.dns_sans) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/net/tls/policy_checking_certificate_verifier.cpp]: const auto &, or is const also automatic ? (multiple places further down as well)

File: vespalib/src/vespa/vespalib/net/tls/tls_context.h [hunk 4]
```diff
@@ -6,11 +6,21 @@
...
+    // Create a TLS context which verifies certificates according to the provided options'
+    // CA trust roots AND allowed peer policies
     static std::shared_ptr<TlsContext> create_default_context(const TransportSecurityOptions&);
+    // Create a TLS context where the certificate verification callback is explicitly provided.
+    // IMPORTANT: This does NOT verify that the peer satisfies the allowed peer policies!
+    //            It only verifies that a peer is signed by a trusted CA. This function should
+    //            therefore only be used in very special circumstances, such as unit tests.
+    static std::shared_ptr<TlsContext> create_default_context(
+            const TransportSecurityOptions&,
+            std::shared_ptr<CertificateVerificationCallback> cert_verify_callback);
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/net/tls/tls_context.h]: alternative solution: simple AND between the custom verifier and the one created from the allowed_peers specification.

File: vespalib/src/vespa/vespalib/net/tls/transport_security_options_reading.cpp [hunk 5]
```diff
@@ -64,24 +70,70 @@ vespalib::string load_file_referenced_by_field(const Cursor& cursor, const char*
     return File::readAll(file_path);
 }
 
+RequiredPeerCredential parse_peer_credential(const Cursor& req_entry) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/net/tls/transport_security_options_reading.cpp]: consider using 'Inspector' for parsing

**Review by havardpe — APPROVED** [review]
👍

**Review by bjorncs — APPROVED** [review]

File: vespalib/src/tests/net/tls/policy_checking_certificate_verifier/policy_checking_certificate_verifier_test.cpp (hunk 2 above)
Comment (vekterli) [reply vespalib/src/tests/net/tls/policy_checking_certificate_verifier/policy_checking_certificate_verifier_test.cpp]: 👀 will fix, should be "authenticated" for both

File: vespalib/src/vespa/vespalib/net/tls/policy_checking_certificate_verifier.cpp (hunk 3 above)
Comment (vekterli) [reply vespalib/src/vespa/vespalib/net/tls/policy_checking_certificate_verifier.cpp]: `const`ness is auto-deduced, but explicitly having it `const` is usually better. Will add.

File: vespalib/src/vespa/vespalib/net/tls/tls_context.h (hunk 4 above)
Comment (vekterli) [reply vespalib/src/vespa/vespalib/net/tls/tls_context.h]: Right now this is only used by tests that explicitly want to control the callback, but if we want to make this functionality more general I agree that a composite callback is the way to go. Will consider for a follow-up.

File: vespalib/src/vespa/vespalib/net/tls/transport_security_options_reading.cpp (hunk 5 above)
Comment (vekterli) [reply vespalib/src/vespa/vespalib/net/tls/transport_security_options_reading.cpp]: Done

**Comment by vekterli** [comment]
Travis fails on
```
Downloading all dependencies. This may take a few minutes with an empty Maven cache.
./bootstrap.sh: line 45: mvn: command not found
./bootstrap.sh: line 45: mvn: command not found
./bootstrap.sh: line 45: mvn: command not found
Building Vespa Maven plugins.
./bootstrap.sh: line 45: mvn: command not found
/source/travis/travis-build-full.sh: line 17: mvn: command not found
```
... I don't think that was me!

**Pull request merged**